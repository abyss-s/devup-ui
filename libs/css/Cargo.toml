[package]
name = "css"
version = "0.1.0"
edition = "2024"

[dependencies]
once_cell = "1.21.3"
phf = { version = "0.12", features = ["macros"] }
serial_test = "3.2.0"
serde = { version = "1.0.219", features = ["derive"] }
regex = "1.11.1"
