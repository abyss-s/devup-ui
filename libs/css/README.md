## Css

Convert css properties to a string.
