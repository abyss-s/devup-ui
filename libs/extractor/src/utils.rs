/// Convert a value to a pixel value
pub fn convert_value(value: &str) -> String {
    let value = value.to_string();
    if let Ok(num) = value.parse::<f64>() {
        let num = num * 4.0;
        return format!("{}px", num);
    }
    value
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_convert_value() {
        assert_eq!(convert_value("1px"), "1px");
        assert_eq!(convert_value("1%"), "1%");
        assert_eq!(convert_value("foo"), "foo");
        assert_eq!(convert_value("4"), "16px");
    }
}
