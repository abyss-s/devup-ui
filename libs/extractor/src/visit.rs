use crate::component::ExportVariableKind;
use crate::extract_style::{ExtractCss, ExtractStyleProperty};
use crate::prop_modify_utils::{modify_prop_object, modify_props};
use crate::style_extractor::{
    extract_style_from_expression, extract_style_from_jsx_attr, ExtractResult,
};
use crate::{ExtractStyleProp, ExtractStyleValue, StyleProperty};
use oxc_allocator::{Allocator, CloneIn};
use oxc_ast::ast::ImportDeclarationSpecifier::ImportSpecifier;
use oxc_ast::ast::JSXAttributeItem::Attribute;
use oxc_ast::ast::JSXAttributeName::Identifier;
use oxc_ast::ast::{
    Argument, BindingPatternKind, CallExpression, Expression, ImportDeclaration,
    ImportOrExportKind, JSXElement, JSXElementName, ObjectPropertyKind, Program, PropertyKey,
    Statement, TaggedTemplateExpression, TemplateElementValue, VariableDeclarator, WithClause,
};
use oxc_ast::visit::walk_mut::{
    walk_call_expression, walk_import_declaration, walk_jsx_element, walk_program,
    walk_tagged_template_expression, walk_variable_declarator,
};
use oxc_ast::{AstBuilder, VisitMut};
use oxc_span::SPAN;
use std::collections::HashMap;

pub struct DevupVisitor<'a> {
    pub ast: AstBuilder<'a>,
    imports: HashMap<String, ExportVariableKind>,
    import_object: Option<String>,
    jsx_imports: HashMap<String, String>,
    css_imports: HashMap<String, String>,
    jsx_object: Option<String>,
    package: String,
    css_file: String,
    pub styles: Vec<ExtractStyleValue>,
}

impl<'a> DevupVisitor<'a> {
    pub fn new(allocator: &'a Allocator, package: &str, css_file: &str) -> Self {
        Self {
            ast: AstBuilder::new(allocator),
            imports: HashMap::new(),
            jsx_imports: HashMap::new(),
            package: package.to_string(),
            css_file: css_file.to_string(),
            styles: vec![],
            import_object: None,
            jsx_object: None,
            css_imports: HashMap::new(),
        }
    }
}

impl<'a> VisitMut<'a> for DevupVisitor<'a> {
    fn visit_program(&mut self, it: &mut Program<'a>) {
        walk_program(self, it);
        if !self.styles.is_empty() {
            it.body.insert(
                0,
                Statement::ImportDeclaration(
                    self.ast.alloc_import_declaration::<Option<WithClause>>(
                        SPAN,
                        None,
                        self.ast
                            .string_literal(SPAN, self.css_file.to_string(), None),
                        None,
                        None,
                        ImportOrExportKind::Value,
                    ),
                ),
            );
        }

        for i in (0..it.body.len()).rev() {
            if let Statement::ImportDeclaration(decl) = &it.body[i] {
                if decl.source.value == self.package && decl.specifiers.iter().all(|s| s.is_empty())
                {
                    it.body.remove(i);
                }
            }
        }
    }
    fn visit_call_expression(&mut self, it: &mut CallExpression<'a>) {
        let jsx = if let Expression::Identifier(ident) = &it.callee {
            self.jsx_imports
                .get(&ident.name.to_string())
                .map(|s| s.to_string())
        } else if let Expression::StaticMemberExpression(member) = &it.callee {
            if let Expression::Identifier(ident) = &member.object {
                if self.jsx_object == Some(ident.name.to_string()) {
                    Some(member.property.to_string())
                } else {
                    None
                }
            } else {
                None
            }
        } else {
            None
        };
        if let Some(j) = jsx {
            if (j == "jsx" || j == "jsxs") && it.arguments.len() > 0 {
                let expr = it.arguments[0].to_expression();
                let element_kind = if let Expression::Identifier(ident) = expr {
                    self.imports.get(ident.name.as_str()).cloned()
                } else if let Expression::StaticMemberExpression(member) = expr {
                    if let Expression::Identifier(ident) = &member.object {
                        if self.import_object == Some(ident.name.to_string()) {
                            ExportVariableKind::try_from(member.property.name.to_string()).ok()
                        } else {
                            None
                        }
                    } else {
                        None
                    }
                } else {
                    None
                };
                if let Some(kind) = element_kind {
                    if it.arguments.len() > 1 {
                        if let Expression::ObjectExpression(obj) =
                            it.arguments[1].to_expression_mut()
                        {
                            let mut tag = kind.to_tag().unwrap_or("div").to_string();
                            let mut props_styles = vec![];
                            for idx in (0..obj.properties.len()).rev() {
                                let mut prop = obj.properties.remove(idx);
                                let mut rm = false;
                                if let ObjectPropertyKind::ObjectProperty(prop) = &mut prop {
                                    if let PropertyKey::StaticIdentifier(ident) = &prop.key {
                                        let name = ident.name.to_string();
                                        rm = match extract_style_from_expression(
                                            &self.ast,
                                            Some(&name),
                                            &mut prop.value,
                                            0,
                                            None,
                                        ) {
                                            ExtractResult::Maintain => false,
                                            ExtractResult::Remove => true,
                                            ExtractResult::ExtractStyle(mut styles) => {
                                                styles.reverse();
                                                props_styles.append(&mut styles);
                                                true
                                            }
                                            ExtractResult::ChangeTag(t) => {
                                                tag = t;
                                                true
                                            }
                                        }
                                    }
                                }
                                if !rm {
                                    obj.properties.insert(idx, prop);
                                }
                            }

                            for ex in kind.extract().into_iter().rev() {
                                props_styles.push(ExtractStyleProp::Static(ex));
                            }

                            for style in props_styles.iter().rev() {
                                self.styles.append(&mut style.extract());
                            }

                            modify_prop_object(&self.ast, &mut obj.properties, props_styles);
                            it.arguments[0] =
                                Argument::StringLiteral(self.ast.alloc_string_literal(
                                    SPAN,
                                    self.ast.atom(tag.as_str()),
                                    None,
                                ));
                        }
                    }
                }
            }
        }
        println!("{:?} {:?}", it, self.css_imports);
        if let Expression::Identifier(ident) = &it.callee {
            if self.css_imports.contains_key(ident.name.as_str()) && it.arguments.len() == 1 {
                if let Argument::ObjectExpression(ref mut obj) = it.arguments[0] {
                    let mut props_styles = vec![];
                    for idx in (0..obj.properties.len()).rev() {
                        let mut prop = obj.properties.remove(idx);
                        let mut rm = false;
                        if let ObjectPropertyKind::ObjectProperty(prop) = &mut prop {
                            if let PropertyKey::StaticIdentifier(ident) = &prop.key {
                                let name = ident.name.to_string();
                                rm = match extract_style_from_expression(
                                    &self.ast,
                                    Some(&name),
                                    &mut prop.value,
                                    0,
                                    None,
                                ) {
                                    ExtractResult::Maintain => false,
                                    ExtractResult::Remove => true,
                                    ExtractResult::ExtractStyle(mut styles) => {
                                        styles.reverse();
                                        props_styles.append(&mut styles);
                                        true
                                    }
                                    ExtractResult::ChangeTag(_) => true,
                                }
                            }
                        }
                        if !rm {
                            obj.properties.insert(idx, prop);
                        }
                    }
                    let mut styles = props_styles
                        .into_iter()
                        .flat_map(|ex| ex.extract())
                        .collect::<Vec<_>>();
                    let class_name = styles
                        .iter()
                        .filter_map(|ex| match ex {
                            ExtractStyleValue::Static(css) => {
                                if let StyleProperty::ClassName(cls) = css.extract() {
                                    Some(cls.to_string())
                                } else {
                                    None
                                }
                            }
                            _ => None,
                        })
                        .collect::<Vec<_>>()
                        .join(" ");

                    self.styles.append(&mut styles);
                    it.arguments[0] = Argument::StringLiteral(self.ast.alloc_string_literal(
                        SPAN,
                        self.ast.atom(&class_name),
                        None,
                    ));
                }
            }
        }
        walk_call_expression(self, it);
    }
    fn visit_tagged_template_expression(&mut self, it: &mut TaggedTemplateExpression<'a>) {
        if let Expression::Identifier(ident) = &it.tag {
            if !self.css_imports.contains_key(ident.name.as_str()) {
                walk_tagged_template_expression(self, it);
                return;
            }

            let css_str = it
                .quasi
                .quasis
                .iter()
                .map(|quasi| quasi.value.raw.as_str())
                .collect::<Vec<&str>>()
                .join("");
            let css = ExtractStyleValue::Css(ExtractCss {
                css: css_str.trim().to_string(),
            });

            if let StyleProperty::ClassName(cls) = css.extract() {
                let mut ve = oxc_allocator::Vec::new_in(self.ast.allocator);
                ve.push(self.ast.template_element(
                    SPAN,
                    false,
                    TemplateElementValue {
                        cooked: None,
                        raw: self.ast.atom(cls.as_str()),
                    },
                ));
                it.quasi.quasis = ve;
            }
            self.styles.push(css);
            return;
        }
        walk_tagged_template_expression(self, it);
    }
    fn visit_jsx_element(&mut self, elem: &mut JSXElement<'a>) {
        walk_jsx_element(self, elem);
        // after run to convert css literal

        let opening_element = &mut elem.opening_element;
        let component_name = &opening_element.name.to_string();
        if let Some(kind) = self.imports.get(component_name) {
            let attrs = &mut opening_element.attributes;
            let mut tag_name = kind.to_tag().unwrap_or("div").to_string();
            let mut props_styles = vec![];

            // extract ExtractStyleProp and remain style and class name, just extract
            for i in (0..attrs.len()).rev() {
                let mut attr = attrs.remove(i);
                let mut rm = false;
                if let Attribute(ref mut attr) = &mut attr {
                    if let Identifier(name) = &attr.name {
                        let name = name.to_string();
                        if let Some(at) = &mut attr.value {
                            rm = match extract_style_from_jsx_attr(&self.ast, &name, at, None) {
                                ExtractResult::Maintain => false,
                                ExtractResult::Remove => true,
                                ExtractResult::ExtractStyle(mut styles) => {
                                    styles.reverse();
                                    props_styles.append(&mut styles);
                                    true
                                }
                                ExtractResult::ChangeTag(tag) => {
                                    tag_name = tag;
                                    true
                                }
                            }
                        }
                    }
                }
                if !rm {
                    attrs.insert(i, attr);
                }
            }

            for ex in kind.extract().into_iter().rev() {
                props_styles.push(ExtractStyleProp::Static(ex));
            }
            for style in props_styles.iter().rev() {
                self.styles.append(&mut style.extract());
            }
            // modify!!
            modify_props(&self.ast, attrs, props_styles);

            // change tag name
            let ident = JSXElementName::Identifier(self.ast.alloc_jsx_identifier(SPAN, tag_name));
            opening_element.name = ident.clone_in(self.ast.allocator);
            if let Some(el) = &mut elem.closing_element {
                el.name = ident;
            }
        }
    }
    fn visit_variable_declarator(&mut self, it: &mut VariableDeclarator<'a>) {
        if let Some(Expression::CallExpression(call)) = &it.init {
            if call.arguments.len() == 1 {
                if let (Expression::Identifier(ident), Argument::StringLiteral(arg)) =
                    (&call.callee, &call.arguments[0])
                {
                    if ident.name == "require" {
                        if arg.value == "react/jsx-runtime" {
                            if let BindingPatternKind::BindingIdentifier(ident) = &it.id.kind {
                                self.jsx_object = Some(ident.name.to_string());
                            } else if let BindingPatternKind::ObjectPattern(object) = &it.id.kind {
                                for prop in &object.properties {
                                    if let PropertyKey::StaticIdentifier(ident) = &prop.key {
                                        if let Some(k) = prop
                                            .value
                                            .get_binding_identifier()
                                            .map(|id| id.name.to_string())
                                        {
                                            self.jsx_imports.insert(k, ident.name.to_string());
                                        }
                                    }
                                }
                            }
                        } else if arg.value == self.package {
                            if let BindingPatternKind::BindingIdentifier(ident) = &it.id.kind {
                                self.import_object = Some(ident.name.to_string());
                            } else if let BindingPatternKind::ObjectPattern(object) = &it.id.kind {
                                for prop in &object.properties {
                                    if let PropertyKey::StaticIdentifier(ident) = &prop.key {
                                        if let Ok(kind) = ExportVariableKind::try_from(
                                            prop.value
                                                .get_binding_identifier()
                                                .map(|id| id.name.to_string())
                                                .unwrap_or("".to_string()),
                                        ) {
                                            self.imports.insert(ident.name.to_string(), kind);
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }

        walk_variable_declarator(self, it);
    }
    fn visit_import_declaration(&mut self, it: &mut ImportDeclaration<'a>) {
        walk_import_declaration(self, it);
        if it.source.value != self.package && it.source.value == "react/jsx-runtime" {
            if let Some(specifiers) = &it.specifiers {
                for specifier in specifiers {
                    if let ImportSpecifier(import) = specifier {
                        self.jsx_imports
                            .insert(import.local.to_string(), import.imported.to_string());
                    }
                }
            }

            return;
        }
        if let Some(specifiers) = &mut it.specifiers {
            for i in (0..specifiers.len()).rev() {
                if let ImportSpecifier(import) = &specifiers[i] {
                    if let Ok(kind) = ExportVariableKind::try_from(import.imported.to_string()) {
                        self.imports.insert(import.local.to_string(), kind);

                        // remove specifier
                        specifiers.remove(i);
                    } else if import.imported.to_string() == "css" {
                        self.css_imports
                            .insert(import.local.to_string(), it.source.value.to_string());
                    }
                }
            }
        }
    }
}
