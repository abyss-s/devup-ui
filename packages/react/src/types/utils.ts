export type Merge<T, U> = Omit<T, Extract<keyof T, keyof U>> & U
