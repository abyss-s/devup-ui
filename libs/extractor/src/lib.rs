mod component;
mod gen_class_name;
mod gen_style;
mod media_prop_extract_utils;
mod prop_extract_utils;
mod prop_modify_utils;
mod utils;
mod visit;

use oxc_codegen::Codegen;

use crate::utils::convert_value;
use crate::visit::DevupVisitor;
use css::{sheet_to_classname, sheet_to_variable_name};
use oxc_allocator::Allocator;
use oxc_ast::ast::Expression;
use oxc_ast::VisitMut;
use oxc_parser::{Parser, ParserReturn};
use oxc_span::SourceType;
use std::error::Error;
use std::hash::{DefaultHasher, Hasher};

/// result of extracting style properties from props
#[derive(Debug)]
pub enum ExtractStyleProp<'a> {
    Static(ExtractStyleValue),
    Responsive(Vec<ExtractStyleProp<'a>>),
    /// static + static ex) margin={test?"4px":"8px"} --> className={test?"margin-4px-0":"margin-8px-0"}
    /// static + dynamic ex) margin={test?a:"8px"} --> className={test?"margin-0":"margin-8px-0"} style={{ "--margin-0": a }}
    /// dynamic + dynamic ex) margin={test?a:b} --> className="margin-0" style={{ "--margin-0": test?a:b }}
    /// issue case: dynamic + dynamic ex) margin={test?a:(b ? "8px": c)} --> className="margin-0" style={{ "--margin-0": test?a:(b ? "8px": c) }}
    Conditional {
        condition: Expression<'a>,
        consequent: Option<Box<ExtractStyleProp<'a>>>,
        alternate: Option<Box<ExtractStyleProp<'a>>>,
    },
}
impl ExtractStyleProp<'_> {
    pub fn extract(&self) -> Vec<ExtractStyleValue> {
        match self {
            ExtractStyleProp::Static(style) => vec![style.clone()],
            ExtractStyleProp::Conditional {
                consequent,
                alternate,
                ..
            } => {
                let mut styles = vec![];
                if let Some(consequent) = consequent {
                    styles.append(&mut consequent.extract());
                }
                if let Some(alternate) = alternate {
                    styles.append(&mut alternate.extract());
                }
                styles
            }
            ExtractStyleProp::Responsive(ref array) => {
                array.iter().flat_map(|s| s.extract()).collect()
            }
        }
    }
}
/// Style property for props
pub enum StyleProperty {
    ClassName(String),
    Variable {
        class_name: String,
        variable_name: String,
        identifier: String,
    },
}
pub trait ExtractStyleProperty {
    /// extract style properties
    fn extract(&self) -> StyleProperty;
}

#[derive(Debug, PartialEq, Clone)]
pub struct ExtractStaticStyle {
    /// property
    pub property: String,
    /// fixed value
    pub value: String,
    /// responsive level
    pub level: u8,
    /// selector
    pub selector: Option<String>,
}

impl ExtractStyleProperty for ExtractStaticStyle {
    fn extract(&self) -> StyleProperty {
        StyleProperty::ClassName(sheet_to_classname(
            self.property.as_str(),
            self.level,
            Some(convert_value(self.value.as_str()).as_str()),
        ))
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct ExtractCss {
    /// css code
    pub css: String,
}

impl ExtractStyleProperty for ExtractCss {
    /// hashing css code to class name
    fn extract(&self) -> StyleProperty {
        let mut hasher = DefaultHasher::new();
        hasher.write(self.css.as_bytes());
        StyleProperty::ClassName(format!("d{}", hasher.finish()))
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct ExtractDynamicStyle {
    /// property
    pub property: String,
    /// responsive
    pub level: u8,
    identifier: String,

    /// selector
    pub selector: Option<String>,
}

impl ExtractStyleProperty for ExtractDynamicStyle {
    fn extract(&self) -> StyleProperty {
        StyleProperty::Variable {
            class_name: sheet_to_classname(self.property.as_str(), self.level, None),
            variable_name: sheet_to_variable_name(self.property.as_str(), self.level),
            identifier: self.identifier.clone(),
        }
    }
}

#[derive(Debug, PartialEq, Clone)]
pub enum ExtractStyleValue {
    Static(ExtractStaticStyle),
    Dynamic(ExtractDynamicStyle),
    Css(ExtractCss),
}

impl ExtractStyleValue {
    pub fn extract(&self) -> StyleProperty {
        match self {
            ExtractStyleValue::Static(style) => style.extract(),
            ExtractStyleValue::Dynamic(style) => style.extract(),
            ExtractStyleValue::Css(css) => css.extract(),
        }
    }
}

#[derive(Debug, PartialEq)]
pub struct ExtractOutput {
    // used styles
    pub styles: Vec<ExtractStyleValue>,

    // output source
    pub code: String,
}

pub struct ExtractOption {
    pub package: String,
    pub css_file: Option<String>,
}

pub fn extract(
    filename: &str,
    code: &str,
    option: ExtractOption,
) -> Result<ExtractOutput, Box<dyn Error>> {
    if !code.contains(option.package.as_str()) {
        // skip if not using package
        return Ok(ExtractOutput {
            styles: vec![],
            code: code.to_string(),
        });
    }
    let allocator = Allocator::default();
    let source_type = SourceType::from_path(filename)?;

    let ParserReturn {
        mut program, // AST
        errors,      // Syntax errors
        panicked,    // Parser encountered an error it couldn't recover from
        ..
    } = Parser::new(&allocator, code, source_type).parse();
    if panicked {
        return Err("Parser panicked".into());
    }
    if !errors.is_empty() {
        return Err(errors
            .iter()
            .map(|e| e.to_string())
            .collect::<Vec<_>>()
            .join("\n")
            .into());
    }
    let mut visitor = DevupVisitor::new(
        &allocator,
        &option.package,
        option
            .css_file
            .unwrap_or(format!("{}/devup-ui.css", option.package))
            .as_str(),
    );
    visitor.visit_program(&mut program);

    Ok(ExtractOutput {
        styles: visitor.styles,
        code: Codegen::new().build(&program).code,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;
    use std::hash::{DefaultHasher, Hasher};

    #[test]
    fn extract_just_tsx() {
        assert_debug_snapshot!(extract(
            "test.tsx",
            "const a = 1;",
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            },
        )
        .unwrap());

        assert_debug_snapshot!(extract(
            "test.tsx",
            "<Box gap={1} />",
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            },
        )
        .unwrap());
    }
    #[test]
    fn ignore_special_props() {
        assert_debug_snapshot!(extract(
            "test.tsx",
            r"import {Box} from '@devup-ui/core'
        <Box padding={1} ref={ref} data-test={1} role={2} children={[]} onClick={()=>{}} aria-valuenow={24} key={2} />
        ",
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());
    }
    #[test]
    fn extract_style_props() {
        assert_debug_snapshot!(extract(
            "test.tsx",
            r"import {Box} from '@devup-ui/core'
        <Box padding={1} margin={2} wrong={} />
        ",
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());
        assert_debug_snapshot!(extract(
            "test.tsx",
            r"import {Box as C} from '@devup-ui/core'
                <C padding={1} margin={2} />
                ",
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());
    }

    #[test]
    fn extract_style_props_with_class_name() {
        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import {Box as C} from '@devup-ui/core'
        <C padding={1} margin={2} className="exists class name" />
        "#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());

        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import {Box as C} from '@devup-ui/core'
        <C padding={1} margin={2} className="  exists class name  " />
        "#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());

        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import {Box as C} from '@devup-ui/core'
        <C padding={1} margin={2} className={"exists class name"} />
        "#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());

        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import {Box as C} from '@devup-ui/core'
        <C padding={1} margin={2} className={} />
        "#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());
        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import {Box as C} from '@devup-ui/core'
        <C padding={1} margin={2} className={"a"+"b"} />
        "#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());
    }

    #[test]
    fn extract_class_name_from_component() {
        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import {VStack as C} from '@devup-ui/core'
        <C padding={1} margin={2} className={"a"+"b"} />
        "#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap(),);
    }
    #[test]
    fn extract_responsive_style_props() {
        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box padding={[null,1]} margin={[2,null,4]} />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());
    }

    #[test]
    fn extract_dynamic_style_props() {
        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box padding={someStyleVar} margin={someStyleVar2} />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());
    }

    #[test]
    fn extract_dynamic_responsive_style_props() {
        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box padding={[someStyleVar,null,someStyleVar1]} margin={[null,someStyleVar2]} />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());
    }

    #[test]
    fn extract_compound_responsive_style_props() {
        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box padding={[someStyleVar,undefined,someStyleVar1]} margin={[null,someStyleVar2]} bg="red" />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());
    }

    #[test]
    fn extract_wrong_responsive_style_props() {
        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box padding={[NaN,undefined,null]} margin={Infinity} />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());
    }

    #[test]
    fn extract_variable_style_props_with_style() {
        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box margin={a} style={{ key:value }} />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());

        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box margin={a} style={styles} />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());
    }

    #[test]
    fn extract_conditional_style_props() {
        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box margin={a === b ? "4px" : "3px"} />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());

        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box margin={a === b ? c : d} />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());

        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box margin={a === b ? "4px" : d} />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());
    }

    #[test]
    fn extract_responsive_conditional_style_props() {
        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box margin={[null, a === b ? "4px" : "3px"]} />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());

        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
        <Box margin={["6px", a === b ? "4px" : "3px"]} />;
        "#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());

        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box margin={a === b ? c : [d, e, f, "2px"]} />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());

        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box margin={a === b ? c : [d, e, f, x === y ? "4px" : "2px"]} />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());

        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box margin={a === b ? [d, e, f, x === y ? "4px" : "2px"] : c} />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());

        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box margin={a === b ? [d, e, f, x === y ? "4px" : "2px"] : ["1px", "2px", "3px"]} />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());
        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box margin={[null, a === b && "4px"]} />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());

        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box margin={[null, a === b && "4px", c === d ? "5px" : null]} />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());
    }

    #[test]
    fn extract_logical_case() {
        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box margin={a===b && "1px"} />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());

        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box margin={a===b || "1px"} />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());

        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box margin={a ?? "1px"} />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());
    }
    #[test]
    fn extract_responsive_conditional_style_props_with_class_name() {
        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box margin={[null, a === b ? (q > w ? "4px" : "8px") : "3px"]} className={"exists"} />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());

        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { Box } from "@devup-ui/core";
<Box margin={[null, a === b || "4px"]} className={"exists"} />;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());
    }

    #[test]
    fn extract_selector() {
        assert_debug_snapshot!(extract(
            "test.tsx",
            r"import {Box} from '@devup-ui/core'
        <Box _hover={{
          mx: 1
        }} />
        ",
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());
    }

    #[test]
    fn extract_selector_with_responsive() {
        assert_debug_snapshot!(extract(
            "test.tsx",
            r"import {Box} from '@devup-ui/core'
        <Box _hover={{
          mx: [1, 2]
        }} />
        ",
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());

        assert_debug_snapshot!(extract(
            "test.tsx",
            r"import {Box} from '@devup-ui/core'
        <Box _hover={[{
          mx: 10
        },{
          mx: 20
        }]} />
        ",
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());
    }

    #[test]
    fn extract_static_css_class_name_props() {
        let mut hasher = DefaultHasher::new();
        hasher.write("background-color: red;".as_bytes());
        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import { css } from "@devup-ui/core";
<Box className={css`
  background-color: red;
`}/>;
"#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());
    }

    #[test]
    fn extract_static_css_with_theme() {
        assert_debug_snapshot!(extract(
            "test.tsx",
            r#"import {Box} from '@devup-ui/core'
        <Box color="$nice" />
        "#,
            ExtractOption {
                package: "@devup-ui/core".to_string(),
                css_file: None
            }
        )
        .unwrap());
    }
}
