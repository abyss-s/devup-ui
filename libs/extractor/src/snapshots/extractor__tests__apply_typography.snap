---
source: libs/extractor/src/lib.rs
expression: "extract(\"test.tsx\",\nr#\"import {Text} from '@devup-ui/core'\n        <Text typography=\"bold\" />\n        \"#,\nExtractOption\n{ package: \"@devup-ui/core\".to_string(), css_file: None }).unwrap()"
---
ExtractOutput {
    styles: [
        Typography(
            "bold",
        ),
    ],
    code: "import \"@devup-ui/core/devup-ui.css\";\nimport { Text } from \"@devup-ui/core\";\n<span className=\"typo-bold\" />;\n",
}
